FreeRTOS-Plus/Demo/coreHTTP_Windows_Simulator/HTTP_S3_Download_Multithreaded/DemoTasks/S3DownloadMultithreadedHTTPExample.c

```diff
@@ -0,0 +1,1142 @@
+    xRequestInfo.methodLen = httpexampleHTTP_METHOD_GET_LENGTH;
+    xRequestInfo.pPath = pcPath;
+    xRequestInfo.pathLen = xRequestUriLen;
+
+    /* Set the HTTP "Connection" header to "keep-alive" to allow multiple
+     * requests to be sent over the same established TCP connection. This is
+     * done in order to download the file in parts. */
+    xRequestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Get the length of the S3 file. */
+    xStatus = prvGetS3ObjectFileSize( &xRequestInfo );
+
+    /* Wait for the response task to receive and interpret the server response
+     * to the file size request. */
+    while( xFileSize == 0 )
+    {
+        /* Check if any errors in the response task have occurred. */
+        if( prvCheckNotification( &ulNotification, httpexampleHTTP_FAILURE, pdTRUE ) != pdFALSE )
+        {
+            LogError( ( "Request task: Received error notification from response task while waiting for the file size. Exiting task." ) );
+            xStatus = pdFAIL;
+            break;
+        }
+    }
+
+    /* Set the number of bytes to request in each iteration, defined by the user
+     * in democonfigRANGE_REQUEST_LENGTH. */
+    if( xFileSize < democonfigRANGE_REQUEST_LENGTH )
+    {
+        xNumReqBytes = xFileSize;
+    }
+    else
+    {
+        xNumReqBytes = democonfigRANGE_REQUEST_LENGTH;
+    }
+
+    /* Here we continuously add range requests to the request queue (and keep
+     * track of their count, with xResponseCount), until the entire length of
+     * the file has been requested. We keep track of the next starting byte to
+     * download with xCurByte, and increment by xNumReqBytes after each
+     * iteration, until xCurByte has reached xFileSize. */
+    while( ( xStatus != pdFAIL ) && ( xCurByte < xFileSize ) )
+    {
+        /* Add range request to the request queue. */
+        xStatus = prvRequestS3ObjectRange( &xRequestInfo,
+                                           xCurByte,
+                                           xCurByte + xNumReqBytes - 1 );
+
+        /* Update the starting byte for the next iteration.*/
+        xCurByte += xNumReqBytes;
+
+        /* If the number of bytes left to download is less than the
+         * pre-defined constant xNumReqBytes, set xNumReqBytes to equal the
+         * accurate number of remaining bytes left to download. */
+        if( ( xFileSize - xCurByte ) < xNumReqBytes )
+        {
+            xNumReqBytes = xFileSize - xCurByte;
+        }
+
+        /* If the request was successfully enqueued, we expect a
+         * corresponding response. */
+        xResponseCount += 1;
+    }
+
+    /* Clear this task's notifications. */
+    xTaskNotifyStateClear( NULL );
+    ulNotification = ulTaskNotifyValueClear( NULL, ~( 0U ) );
+
+    /* Notify the other tasks of completion. */
+    xTaskNotify( xMainTask, httpexampleREQUEST_TASK_COMPLETION, eSetBits );
+    xTaskNotify( xResponseTask, httpexampleREQUEST_TASK_COMPLETION, eSetBits );
```

Comment:
Is this notification used by the response task?

Reply:
The task completion notifications are currently only checked for in the main task -- since notifying the other tasks is unnecessary in this case, I will remove those lines. Any relevant errors in other tasks would present themselves earlier through the `httpexampleHTTP_FAILURE` notification.